_posts/2015-02-20-elasticsearch-and-denormalization.md

```diff
@@ -0,0 +1,223 @@
+  end
+
+  def display_name
+    sample.display_name
+  end
+
+  def original_image
+    sample.original_image
+  end
+
+  def color_name
+    sample.color.name
+  end
+
+  def style_name
+    style.name
+  end
+
+  def style_id
+    style.id
+  end
+
+  # One of the powers of this pattern is for attributes like this
+  # This method returns a boolean value that is complex and expensive to
+  # calculate
+  # What's more, it is nearly impossible to query with SQL.
+  # Once you put it in the ElasticSearch index it is cheap to store and return
+  # and simple to query
+  def has_basic_attributes?
+    return false unless style_scalar_attributes
+    required_scalar_values     = REQUIRED_BASIC_SCALAR_ATTRIBUTES.map{|attr| style_scalar_attributes.send(attr) }
+    required_multi_attributes  = if root_item_type  == 'Apparel'
+                                   REQUIRED_BASIC_APPAREL_MULTI_ATTRIBUTES
+                                 else
+                                   []
+                                 end
+    required_multi_values      = required_multi_attributes.map do |attr|
+      style_multi_attributes_names.include?(attr)
+    end
+    (required_scalar_values + required_multi_values).all?
+  end
+
+  # etc ...
+
+  private
+  # any methods that aren't included in the hash can be private and should be
+  # memoized for better performance
+  def style
+    @style  ||= sample.style
+  end
+
+end
+```
+
+You can use this with [Elasticsearch::Model][elasticsearch-model] library to
+build the hash for the index:
+
+[elasticsearch-model]: https://github.com/elasticsearch/elasticsearch-rails/tree/master/elasticsearch-model
+
+```ruby
+class Sample < ActiveRecord::Base
+  include Elasticsearch::Model
+
+  def to_hash
+    SampleDenormalizer.new(self).to_hash
+  end
+  
+  # you still have to include the mapping here
+  settings(elasticsearch_settings) do
+    mapping do
+      indexes :display_name, type: 'multi_field', fields: {
+        display_name:       { type: "string", analyzer: "snowball" },
+        display_name_exact: { type: "string", index: :not_analyzed }
+      }
+      # etc ...
+    end
+  end
+end
+```
+
+So that's how you can use the Denormalizer pattern to build up your search
+index. You can use that same code for more typical presenter work.
+
+```ruby
+class SamplesController < ApplicationController
+
+  # the objects being presented here are Hashie::Mash objects from ElasticSearch
+  def index
+    @samples  = Sample.search(params[:keywords]).map{|s| SamplePresenter.new(s._source) }
+  end
+
+  # the object being presented here is an ActiveRecord object
+  # this action is being called by an AJAX request
+  def update
+    sample  = Sample.find(params[:id])
+    sample.update!(params[:sample])
+    @sample = SamplePresenter.new(sample)
+    render "samples/show", layout: false
+  end
+
+end
+```
+
+Both the `Hashie::Mash` and `ActiveRecord` objects are using the same presenter.
```

Comment:
Both the `Hash::Mash` **(which is what the ElasticSearch Ruby client returns)**, and `ActiveRecord`